### ISSUE TITLE
index is out of bounds for dimension X with size Y [No Nan/Inf Data]

### ISSUE THREAD

**Issue by fsilavong** [dreamquark-ai/tabnet#352]
**Describe the bug**

Encountered the error below, similar to #186, but the **data does not contain any nan, inf or -inf.** The error only occurs for some of  hyper-parameter, but not others. 

Very much appreciate any guidance on this!!

**What is the current behavior?**

Encountered error below:

```
index -1 is out of bounds for dimension 1 with size 2911
```
The traceback leads us to:

```
sparsemax.py in _threshold_and_support(input, dim)
74
75 support_size = support.sum(dim=dim).unsqueeze(dim)
---> 76 tau = input_cumsum.gather(dim, support_size - 1)
77 tau /= support_size.to(input.dtype)
78 return tau, support_size
```

The input_cumsum is a vector of all 0 and input is nan. 

**If the current behavior is a bug, please provide the steps to reproduce.**

**Expected behavior**

No errors 

**Screenshots**

**Other relevant information:**
poetry version:  
python version: 3.6
Operating System: Linux
Additional tools:

**Additional context**

**Comment by Optimox** [comment]
Does it solve your problem if you set `drop_last=True` ?

**Comment by fsilavong** [comment]
> Does it solve your problem if you set `drop_last=True` ?

Sadly, not

**Comment by Optimox** [comment]
Then I'll probably need more information in order to help, do you have a minimal reproducible code?

**Comment by Optimox** [comment]
@fsilavong I'm closing this as there is no way to reproduce your error.

Please feel free to reopen when more information is available.

**Comment by BardiaKh** [comment]
I see the same error. I'm using TabNet as nn.module and all of a sudden face this issue in training my network. I curated the validation set so all the batches have the same size.  I get this error in epoch 22 of training, which is odd to me. I use deterministic training so I always get it at the same spot, and always during validation.

I change the learning rate and it happens in another epoch. I even changed the sparsemax to entmax but still get this error. 

I know this information might not be much, but I'd appreciate it if you can guide me in the right direction to fix it.

**Comment by Optimox** [comment]
hello @BardiaKh,

Without a reproducible error I can't really help.

However here is a list of points that might be useful:

- `drop_last=False` ensures that you always have full batches during training, errors can occur if you have one element in the batch because of batch normalization. This should not be a problem for inference so there is no need to curate your validation set.
- can you validate that the error occurs during validation ? Can you train without any error if you put no validation set at all?
- can you make sure that you have no nan, infinite values or mixed values in your dataset (either training or validation)?
- **Probably my first choice** : if things are suddenly going wild at epoch 22 while everything was working fine before this might be caused by a gradient issue. Have you tried to set `clip_value=1` or `clip_value=5` for gradient clipping?

Let us know if any of this solved your problem!
